code/modules/library/library_equipment.dm

```diff
@@ -0,0 +1,412 @@
+	new /obj/item/book/manual/wiki/engineering_guide(src)
+	new /obj/item/book/manual/engineering_singularity_safety(src)
+	new /obj/item/book/manual/wiki/robotics_cyborgs(src)
+	update_icon()
+
+/obj/structure/bookcase/manuals/research_and_development
+	name = "R&D Manuals bookcase"
+
+/obj/structure/bookcase/manuals/research_and_development/Initialize()
+	. = ..()
+	new /obj/item/book/manual/research_and_development(src)
+	update_icon()
+
+/obj/structure/bookcase/sop
+	name = "bookcase (Standard Operating Procedures)"
+
+/obj/structure/bookcase/sop/Initialize()
+	. = ..()
+	new /obj/item/book/manual/wiki/sop_command(src)
+	new /obj/item/book/manual/wiki/sop_engineering(src)
+	new /obj/item/book/manual/wiki/sop_general(src)
+	new /obj/item/book/manual/wiki/sop_legal(src)
+	new /obj/item/book/manual/wiki/sop_medical(src)
+	new /obj/item/book/manual/wiki/sop_science(src)
+	new /obj/item/book/manual/wiki/sop_security(src)
+	new /obj/item/book/manual/wiki/sop_service(src)
+	new /obj/item/book/manual/wiki/sop_supply(src)
+	update_icon()
+
+/obj/structure/bookcase/random
+	var/category = null
+	var/book_count = 5
+	icon_state = "random_bookcase"
+	anchored = TRUE
+
+/obj/structure/bookcase/random/Initialize(mapload)
+	. = ..()
+	var/list/books = GLOB.library_catalog.get_random_book(book_count, doAsync = FALSE)
+	for(var/datum/cachedbook/book as anything in books)
+		new /obj/item/book(src, book, TRUE, FALSE)
+	update_icon()
+
+/*
+ * Book binder
+ */
+/obj/machinery/bookbinder
+	name = "Book Binder"
+	icon = 'icons/obj/library.dmi'
+	icon_state = "binder"
+	anchored = TRUE
+	density = TRUE
+
+	var/datum/cachedbook/selected_content = new()
+	var/printing = FALSE
+
+/obj/machinery/bookbinder/attack_ai(mob/user)
+	return attack_hand(user)
+
+/obj/machinery/bookbinder/attack_hand(mob/user)
+	if(..())
+		return
+	ui_interact(user)
+
+/obj/machinery/bookbinder/attack_ghost(mob/user)
+	ui_interact(user)
+
+/obj/machinery/bookbinder/attackby(obj/item/I, mob/user)
+	if(default_unfasten_wrench(user, I))
+		power_change()
+		return
+	if(istype(I, /obj/item/paper))
+		select_paper(I)
+	if(istype(I, /obj/item/paper_bundle))
+		select_paper_stack(I)
+	if(istype(I, /obj/item/book))
+		select_book(I)
+	else
+		return ..()
+
+/obj/machinery/bookbinder/proc/select_paper(obj/item/paper/P)
+	selected_content.title = P.name
+	selected_content.author = null
+	selected_content.content = list(P.info)
+
+/obj/machinery/bookbinder/proc/select_paper_stack(obj/item/paper_bundle/P)
+	selected_content.title = P.name
+	selected_content.author = null
+	selected_content.content = list()
+	for(var/obj/item/paper/I in P.contents)
+		selected_content.content += I.info
+
+/obj/machinery/bookbinder/proc/select_book(obj/item/book/B)
+	if(!B || B.protected || B.carved)
+		return
+	selected_content.title = B.title
+	selected_content.author = B.author
+	selected_content.author = B.summary
+	selected_content.content = B.pages
+	for(var/c in B.categories)
+		if(c)
```

Comment:
Can a book contain null categories?

Reply:
yes

Reply:
alright so these are the problem lines
```dm
/datum/cachedbook/proc/LoadFromRow(list/row)
	...
	categories += text2num(row["primary_category"])
	categories += text2num(row["secondary_category"])
	categories += text2num(row["tertiary_category"])```
        ...
```
I will have to do a zero value check here, time to see if this functions fine with it

Reply:
did that, seems to work fine, although it won't hurt to keep the check here. The alternative being no check and possible TGUI crashes.